f/common_logic/data_validation.py

```diff
@@ -0,0 +1,152 @@
+import csv
+import json
+import logging
+from pathlib import Path
+
+import pandas as pd
+from lxml import etree
+
+logging.basicConfig(level=logging.INFO)
+logger = logging.getLogger(__name__)
+
+
+def detect_structured_data_type(file_path: str) -> str:
+    """
+    Detects the type of a structured data file, focusing on geospatial and tabular formats.
+
+    Only files with recognized extensions are considered. The function attempts to
+    parse the file to confirm it matches the expected structure. If the file type is
+    unsupported or the content does not validate, 'unsupported' is returned.
+
+    This function is intended to identify structured datasets used in environmental,
+    geographic, and tabular data pipelines. It supports:
+    - Geospatial formats: geojson, kml, gpx
+    - Tabular formats: csv, xls, xlsx
+    - General structured formats: json
+
+    TODO: Add support for ESRI shapefiles
+
+    Parameters
+    ----------
+    file_path : str
+        Path to the file to inspect.
+
+    Returns
+    -------
+    str
+        The validated file type (e.g., "csv", "geojson", etc.), or "unsupported" if
+        the file does not conform to expectations.
+    """
+
+    path = Path(file_path)
+    if not path.exists():
+        raise FileNotFoundError(f"File does not exist: {file_path}")
+
+    # Only allow files with known extensions
+    suffix_map = {
+        ".geojson": "geojson",
+        ".json": "json",
+        ".xlsx": "xlsx",
+        ".xls": "xls",
+        ".gpx": "gpx",
+        ".kml": "kml",
+        ".csv": "csv",
+    }
+
+    suffix = path.suffix.lower()
+
+    logger.debug(f"Checking file {file_path} with suffix {suffix}")
+
+    if suffix not in suffix_map:
+        return "unsupported"
+
+    expected_type = suffix_map[suffix]
+
+    def get_namespace(tag: str) -> str:
+        if tag.startswith("{"):
+            return tag[1:].split("}")[0]
+        return ""
+
+    # Try validators in priority order; first match wins
```

Comment:
I think this comment isn't right. Your fn doesn't make multiple tries (in any order). It tries the one validator associated with the file extension and either it passes or we return "unrecognized".

So for example if you passed a geojson, with `.json` file extension, it would return "json". That's probably fine (it matches "Only files with recognized extensions are considered" from the docstring)

I'm just saying the code reads in a way that doesnt match this comment.  I'd also recommend considering to group `suffix_map` and `validators` into the same variable (for example a dict from suffix to `(name, val_fn)` tuple).  I think it might read more cleanly.

Reply:
Good catch. Comment was leftover from a working idea that was abandoned.